core/src/test/java/org/elasticsearch/cluster/routing/DelayedAllocationServiceTests.java

```diff
@@ -0,0 +1,473 @@
+        // starting primaries
+        clusterState = ClusterState.builder(clusterState)
+            .routingResult(allocationService.applyStartedShards(clusterState, clusterState.getRoutingNodes().shardsWithState(INITIALIZING)))
+            .build();
+        // starting replicas
+        clusterState = ClusterState.builder(clusterState)
+            .routingResult(allocationService.applyStartedShards(clusterState, clusterState.getRoutingNodes().shardsWithState(INITIALIZING)))
+            .build();
+        assertFalse("no shards should be unassigned", clusterState.getRoutingNodes().unassigned().size() > 0);
+        String nodeId = null;
+        final List<ShardRouting> allShards = clusterState.getRoutingNodes().routingTable().allShards("test");
+        // we need to find the node with the replica otherwise we will not reroute
+        for (ShardRouting shardRouting : allShards) {
+            if (shardRouting.primary() == false) {
+                nodeId = shardRouting.currentNodeId();
+                break;
+            }
+        }
+        assertNotNull(nodeId);
+
+        // remove node that has replica and reroute
+        clusterState = ClusterState.builder(clusterState).nodes(DiscoveryNodes.builder(clusterState.nodes()).remove(nodeId)).build();
+        clusterState = ClusterState.builder(clusterState).routingResult(allocationService.reroute(clusterState, "reroute")).build();
+        ClusterState stateWithDelayedShard = clusterState;
+        // make sure the replica is marked as delayed (i.e. not reallocated)
+        assertEquals(1, UnassignedInfo.getNumberOfDelayedUnassigned(stateWithDelayedShard));
+        assertEquals(nodeLeftTimestampNanos,
+            stateWithDelayedShard.getRoutingNodes().unassigned().iterator().next().unassignedInfo().getUnassignedTimeInNanos());
+
+        assertNull(delayedAllocationService.delayedRerouteTask.get());
+        long delayUntilClusterChangeEvent = TimeValue.timeValueNanos(randomInt((int)shorterDelaySetting.nanos() - 1)).nanos();
+        long clusterChangeEventTimestampNanos = nodeLeftTimestampNanos + delayUntilClusterChangeEvent;
+        delayedAllocationService.setNanoTimeOverride(clusterChangeEventTimestampNanos);
+        delayedAllocationService.clusterChanged(new ClusterChangedEvent("fake node left", stateWithDelayedShard, clusterState));
+
+        // check that delayed reroute task was created and registered with the proper settings
+        DelayedAllocationService.DelayedRerouteTask delayedRerouteTask = delayedAllocationService.delayedRerouteTask.get();
+        assertNotNull(delayedRerouteTask);
+        assertFalse(delayedRerouteTask.cancelScheduling.get());
+        assertThat(delayedRerouteTask.baseTimestampNanos, equalTo(clusterChangeEventTimestampNanos));
+        assertThat(delayedRerouteTask.nextDelay.nanos(),
+            equalTo(delaySetting.nanos() - (clusterChangeEventTimestampNanos - nodeLeftTimestampNanos)));
+
+        // update settings with shorter delay
```

Comment:
can we randomly test another similar path where a shard from another index becomes unassinged and have a shorter delay? this should reschedule as well..